rendercanvas/html.py

```diff
@@ -415,39 +425,30 @@ def _rc_present_bitmap(self, **kwargs):
             c = (ctypes.c_uint8 * m.nbytes).from_buffer(data)  # No copy
             array_uint8 = to_js(c)  # Makes a copy, and somehow mangles the data??
             array_uint8_clamped = Uint8ClampedArray.new(array_uint8.buffer)  # no-copy
-
         # Create image data
         image_data = ImageData.new(array_uint8_clamped, w, h)
 
-        # Convert to image bitmap.
-        # AK: I strongly suspect that this is actually a GPU texture. This would explain why creating an ImageBitmap is async.
-        size = self._canvas_element.width, self._canvas_element.height
-        promise = window.createImageBitmap(
-            image_data,
-            {
-                "resizeQuality": "pixelated",
-                "resizeWidth": int(size[0]),
-                "resizeHeight": int(size[1]),
-            },
-        )
-
-        # So now we wait for the ImageBitmap to be ready, and then present it in the animation frame
-
-        @create_proxy
-        def schedule_final_present(image_bitmap):
-            self._pending_image_bitmap = image_bitmap
-            window.requestAnimationFrame(create_proxy(self._present_final))
-
-        promise.then(schedule_final_present)
-
-    def _present_final(self, _timestamp=None):
-        if self._pending_image_bitmap is not None:
-            self._html_context.transferFromImageBitmap(self._pending_image_bitmap)
-            self._pending_image_bitmap.close()
-        self._pending_image_bitmap = None
-        # Maybe schedule a new draw
-        if self._draw_is_requested:
-            loop.call_soon(self._rc_force_draw)
+        # Now present the image data.
+        # For this we can blit the image into the canvas (i.e. no scaling). We can only use this is the image size matches
+        # the canvas size (in physical pixels). Otherwise we have to scale the image. For that we can use an ImageBitmap and
+        # draw that with CanvasRenderingContext2D.drawImage() or ImageBitmapRenderingContext.transferFromImageBitmap(),
+        # but creating an ImageBitmap is async, which complicates things. So we use an offscreen canvas as an in-between step.
+        cw, ch = self._canvas_element.width, self._canvas_element.height
+        if w == cw and h == ch:
+            # Quick blit
+            self._canvas_element.getContext("2d").putImageData(image_data, 0, 0)
```

Comment:
out of curiosity: it is strictly necessary to download the image data to a JS buffer and then push it to a canvas? can't we render to the canvas context directly?

I imagine that's where you get the webgpu context from anyway?

Reply:
This is primarily meant for the "bitmap" context. We could also use it for the wgpu contetx, but the plan is to allow writing to the wgpu context directly.